src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+For the remaining minima the widest bounding box side is at least 3.
+"""
+diameter_closing(image::AbstractArray; kwargs...) =
+    diameter_closing!(similar(image), image; kwargs...)
+
+# Calculates the local maxima or minima of the image using the max-tree
+# representation (maxima if maxtree.rev=false, minima otherwise).
+# It's not the most effecient method for extrema calculation, but could be
+# useful if the max-tree is already calculated.
+# Each minima/maxima region is labeled with the unique id (1 is the global
+# maximum/minimum).
+function local_extrema!(output::AbstractArray{<:Any, N},
+                        image::AbstractArray{<:Any, N},
+                        maxtree::MaxTree{N}) where N
+    (axes(image) == axes(maxtree)) || throw(DimensionMismatch())
+    (size(output) == size(image)) || throw(DimensionMismatch())
+
+    next_label = one(eltype(output))
+    fill!(output, next_label) # initialize output (just has to be non-zero)
+    @inbounds for p in Iterators.reverse(maxtree.traverse)
+        q = maxtree.parents[p]
+        # if p is canonical (parent has a different value)
+        if image[p] != image[q]
+            output[q] = 0
+            # if output[p] was the parent of some other canonical
+            # pixel, it has been set to zero. Only the leaves
+            # (local maxima) are thus > 0.
+            if output[p] > 0
+                output[p] = next_label
+                next_label += 1
+            end
+        end
+    end
+
+    @inbounds for p in Iterators.reverse(maxtree.traverse)
+        q = maxtree.parents[p]
+        # if p is not canonical (parent has the same value)
+        if image[p] == image[q]
+            # in this case we propagate the value
+            output[p] = output[q]
+        end
+    end
+
+    return output
+end
+
+"""
+    local_maxima!(output, image; [connectivity=1], [maxtree=nothing]) -> output
+
+Detects the local maxima of `image` and stores the result in `output`.
+See [`local_maxima`](@ref) for the detailed description of the method.
+"""
+function local_maxima!(output::AbstractArray, image::AbstractArray;
+                       connectivity::Integer=1,
+                       maxtree::Union{MaxTree, Nothing} = nothing)
+    check_output_image(output, image)
+    _maxtree = check_maxtree(maxtree, image, connectivity=connectivity, rev=false)
+    return local_extrema!(output, image, _maxtree)
+end
+
+"""
+    local_maxima(image::AbstractArray; [connectivity=1], [maxtree=nothing]) -> Array
+
+Determines and labels all *local maxima* of the `image`.
+
+# Details
+The *local maximum* is defined as the connected set of pixels that have the
+same value, which is grater then the values of all pixels in direct
+neighborhood of the set.
+
+Technically, the implementation is based on the *max-tree* representation
+of an image. It's efficient if the max-tree representation
+has already been computed. Otherwise, it is preferable to use
+the function `local_maxima`.
```

Comment:
Since this function is called `local_maxima`, this remark is confusing.

Reply:
I've changed it into "alternative method", but if you know dedicated local_min/max operations in the other packages, I can mention them.

Reply:
`findlocalminima` and `findlocalmaxima`

Reply:
I've updated the docs. But we may also consider renaming these functions to `findlocalminima/maxima`, so that the choice of the optimal method will happen on the dispatch level (when `maxtree` is supplied, this version would be used).
If so, how to better organize this?